### PR TITLE
Drop charge value completely

### PR DESCRIPTION
**Pull request by S-S-X**
Made another PR for this to make it cleaner and because new PR was not that expensive (and I wanted to have separate branch locally).
Tests at least passed unmodified but current tests are not covering that much... also link #233 

@OgelGames 

Bonus: this should be also option with best possible performance and minimum complexity. API wont change (from what it is at target PR).

### CODE REVIEW

**Comment by S-S-X** [comment]
Ofc managed to push syntax error and luacheck failure but this would be inaccurate energy storage.
Range guards can be added if needed but I think it is not needed (should try what happens if `stack:set_wear(x)` is called with invalid values.

If it wont crash then I would not add guards but require tools to provide correct numbers.

**Comment by S-S-X** [comment]
Well, seems I got values backwards but basic idea works. Just little issue with using tool charges it and charging in bb discharges 🤣 
Requires few adjustments to calculations but other than that seems fine.

**Comment by S-S-X** [comment]
Oh... actually wear value 0 is no wear at all and 1 is max condition, so that's where problem is actually (in addition to having it backwards). So have to go from 1 to 65535 + 1 for overflow instead of 0-65535.

Somehow I was thinking that 65535 is pristine and towards zero is "broken" tool.

File: technic/helpers.lua [hunk 1]
```diff
@@ -62,44 +62,26 @@ function technic.swap_node(pos, name)
...
-		end
-	end
-end
-
 function technic.set_RE_charge(stack, charge)
-	local meta = stack:get_meta()
-	charge = math.max(0, charge)
-	set_RE_wear(stack, charge)
-	meta:set_int("charge", charge)
+	stack:set_wear( -math.floor(charge * (65535 / stack:get_definition().max_charge) + 0.5) )
```
Comment (S-S-X) [inline technic/helpers.lua:66]: not sure if it is good idea to abuse this... maybe it is fine?

**Comment by S-S-X** [comment]
Maybe I'll still work on this a bit and try to make sure that `technic.use_RE_charge` will always add at least single point to wear. Other than that only "problem" I can see is with extremely high max charge combined with extremely low charge per but have not yet seen tools where this matters.
If there's tool that has over 65535 individual uses from full charge then that one would not work well (it would be limited to 65535 uses).

**Comment by S-S-X** [comment]
Tool charging and discharging rate is high enough to not be likely source of problems. Would be still good to check that input / output error is small enough to not be significant (so that abusing it wont be useful) and maybe even make it always lose charge instead of gaining (so that abusing it would just simulate battery stress and losses).

**Comment by S-S-X** [comment]
I think I'll merge this and continue testing on main branch.

**Upsides:**
* This will solve tool upgrades nicely
* Provides simpler data storage
* Reduced LoC
* Better performance
* Easy way to simulate non ideal charge storage for tools.

**Downsides:**
* Limits simple tool charge use to 65535 (have not seen any tools that do this, high capacity tools I've seen are around 500)

**Comment by OgelGames** [comment]
Yeah, I think the upsides far outweigh that one downside 👍

> try to make sure that `technic.use_RE_charge` will always add at least single point to wear

That can be done simpler; round when getting the charge, but floor when setting it: (tested with high-charge tool)
```lua
function technic.set_RE_charge(stack, charge)
	if charge < 1 then
		stack:set_wear(0)
	else
		local max_charge = stack:get_definition().max_charge
		stack:set_wear(65536 - math.floor(charge * 65535 / max_charge))
	end
end

function technic.get_RE_charge(stack)
	local wear = stack:get_wear()
	if wear < 1 then
		return 0
	else
		local max_charge = stack:get_definition().max_charge
		return math.floor(((65536 - wear) * max_charge / 65535) + 0.5)
	end
end

function technic.use_RE_charge(stack, amount)
	if technic.creative_mode or amount < 1 then
		-- Do not check charge in creative mode or when trying to use zero amount
		return true
	end
	local charge = technic.get_RE_charge(stack)
	if charge < amount then
		-- Not enough energy available
		return false
	end
	technic.set_RE_charge(stack, charge - amount)
	-- Charge used successfully
	return true
end
```

**Comment by S-S-X** [comment]
> Yeah, I think the upsides far outweigh that one downside 👍
> 
> > try to make sure that `technic.use_RE_charge` will always add at least single point to wear
> 
> That can be done simpler; round when getting the charge, but floor when setting it: (tested with high-charge tool)
> 
> ```lua
> function technic.set_RE_charge(stack, charge)
> 	if charge < 1 then
> 		stack:set_wear(0)
> 	else
> 		local max_charge = stack:get_definition().max_charge
> 		stack:set_wear(65536 - math.floor(charge * 65535 / max_charge))
> 	end
> end
> 
> function technic.get_RE_charge(stack)
> 	local wear = stack:get_wear()
> 	if wear < 1 then
> 		return 0
> 	else
> 		local max_charge = stack:get_definition().max_charge
> 		return math.floor(((65536 - wear) * max_charge / 65535) + 0.5)
> 	end
> end
> 
> function technic.use_RE_charge(stack, amount)
> 	if technic.creative_mode or amount < 1 then
> 		-- Do not check charge in creative mode or when trying to use zero amount
> 		return true
> 	end
> 	local charge = technic.get_RE_charge(stack)
> 	if charge < amount then
> 		-- Not enough energy available
> 		return false
> 	end
> 	technic.set_RE_charge(stack, charge - amount)
> 	-- Charge used successfully
> 	return true
> end
> ```

This seems to make it way less accurate than using custom floating point math:

Charging 6 rounds, 10kEU / cycle. Result: error is way higher.
```
Failure → spec/tools_spec.lua @ 201
Technic power tool charging and discharging extreme ratios t1_65536 can be charged
spec/tools_spec.lua:207: Expected objects to be equal.
Passed in: (number) 59997 Expected: (number) 60000
```

Charging 6 rounds, 10kEU / cycle. Result: error is way higher.
```
Failure → spec/tools_spec.lua @ 223
Technic power tool charging and discharging extreme ratios t100_6553600 can be charged
spec/tools_spec.lua:231: Expected (number) 69998 to be less than (number) 69501
```

Using extreme ratio tools, 7 rounds, 100 per use, initial charge 700, expected after 6 round: 100, expected after 7 rounds 0. Result: stack destroyed.
```
Failure → spec/tools_spec.lua @ 235
Technic power tool charging and discharging extreme ratios t100_6553600 can be used
...ampos/.luarocks/share/lua/5.1/mineunit/itemstack.lua:29: ItemStack:set_wear invalid wear value: 65536
```

**Comment by S-S-X** [comment]
These tools registered for tests have extreme max_charge to use ratios (and use values that are likely to cause errors with rounding) like `mymod:t100_6553600` has `max_charge = 6553600` and calls `technic.use_RE_charge(stack, 100)`.
And `mymod:t1_65536` has `max_charge = 65536` and calls `technic.use_RE_charge(stack, 1)`.

**Comment by S-S-X** [comment]
That Chainsaw MK3 from technic_addons actually also somewhat works when cutting large enough area, anything more than just few nodes.

21 x 11 x 21 area actual charge used 48508 where accurate use would be 48510 (10 per node).

**Comment by OgelGames** [comment]
Ah, I see the problem with my code, it floors the value when charging, but it should only floor it when using it, and round when charging.

**Comment by S-S-X** [comment]
Also `stack:set_wear(65536 - math.floor(charge * 65535 / max_charge))` sets invalid value for charge = 1.
But I've not found problems with current code so unless you got something some a lot simpler solution then I guess current is what should be used.

**Comment by S-S-X** [comment]
Well, cleaned up negative wear value hack (while kind of nice not sure if it is really safe).
Also made it calculate conversion factor during registration and use it instead of max_charge.

**Comment by S-S-X** [comment]
Maybe `max_charge` should also be `technic_max_charge` instead...

**Comment by OgelGames** [comment]
> But I've not found problems with current code so unless you got something some a lot simpler solution then I guess current is what should be used.

There is one problem, `set_RE_charge` doesn't do the wear check. I think it should be moved out of `use_RE_charge` and into `set_RE_charge`.

This code *should* work for that: (idk how to run tests locally)
```lua
function technic.set_RE_charge(stack, charge)
	local wear = 65536 - charge * stack:get_definition().technic_wear_factor
	stack:set_wear(wear > stack:get_wear() and math.ceil(wear) or math.floor(wear + 0.5))
end
```

**Comment by S-S-X** [comment]
For _"how to run tests locally"_, I'm pretty sure Mineunit is not compatible with Windows so either Windows 10 WSL should be used or some virtual machine running some Linux distro.
From command line you'd just install Mineunit from Luarocks: `luarocks install --server=https://luarocks.org/dev mineunit` and run it in mod source directory with: `cd mods/technic/technic` and `mineunit`

**Comment by S-S-X** [comment]
> I should add few more tests that especially try to abuse charging to gain more than what was pushed from BB as current set does not do that as much.

Yeah, just what I thought. Error in these cases generates small amounts of free energy, however that error is way smaller which indicates that idea might be correct but calculations done with inaccurate values rounded too early and used in calculations after rounding.

**Comment by OgelGames** [comment]
> What do you mean by "wear check"? Always rounding and possibly getting "over charged"?

This: https://github.com/mt-mods/technic/blob/d51c4dd8ff60d38d16f7c0d49d41d8ec0cc366ac/technic/helpers.lua#L85-L88

Will all these errors I'm wondering if maybe this isn't such a good idea after all... metadata values seem so much better now...

**Comment by S-S-X** [comment]
Well, of course error is larger because it tries to always go into one direction. I think that's probably not significant because error is similar for both charging and discharging.
Probably real reason to do it would be "battery condition simulation" I mentioned which would be bit more complicated anyway and would like to have required maintenance for tools.

**Comment by S-S-X** [comment]
> > What do you mean by "wear check"? Always rounding and possibly getting "over charged"?
> 
> This:
> 
> https://github.com/mt-mods/technic/blob/d51c4dd8ff60d38d16f7c0d49d41d8ec0cc366ac/technic/helpers.lua#L85-L88
> 
> Will all these errors I'm wondering if maybe this isn't such a good idea after all... metadata values seem so much better now...

I don't see any significant errors there, could you give some example data to make it easier to understand?

**Comment by OgelGames** [comment]
> I don't see any significant errors there, could you give some example data to make it easier to understand?

I mean the errors/inaccuracies in charge values, not errors in the code.

**Comment by S-S-X** [comment]
So far I've only seen errors increase after testing code suggested in comments while with original code I've only seen errors (which are smaller) balancing out between charge / discharge cycles.

**Comment by S-S-X** [comment]
I think it would be possible to increase accuracy further with error correction by interpolation based on before/after error difference.
Basically if there's error then do linear interpolation for errors wear before/wear after and remove that from value calculated with inaccurate scaled values.

Does this sound about right? This is based on assumption that last change was same as current (_which is true for charging and discharging for all but last cycle_). Also it is of course balancing instead of direct correction because there simply is not enough data for direct correction.

Also I have not really tested current code in game yet so not sure if it is even needed but probably good to have note here at least.

**Comment by BuckarooBanzay** [comment]
I merged this PR into the test-server and did some tests with various power-tools (using, discharging, recharging)
everything works as expected :+1:

**Review by BuckarooBanzay — APPROVED** [review]

**Comment by S-S-X** [comment]
Tested charging / discharging tools in loop, for me it seems to be very good. More comments on Discord https://discord.com/channels/513329453741637637/816367895621795861/909717067677175808

It seems like even if there's probably some combinations with exact values it wont be enough to make batterybox + powertool power plants viable, it is simply too slow and it should be easy to if needed because having only wear value leaves all upgrade options open without additional cleanup or leftover data.

**Pull request merged**